File: src/main/java/cn/wizzer/jellyfin/pinyin/JellyfinHandler.java
package cn.wizzer.jellyfin.pinyin;

import cn.wizzer.jellyfin.pinyin.utils.JellyfinUtil;
import cn.wizzer.jellyfin.pinyin.utils.PinyinUtil;
import org.nutz.ioc.loader.annotation.IocBean;
import org.nutz.json.Json;
import org.nutz.lang.Strings;
import org.nutz.lang.Tasks;
import org.nutz.lang.util.NutMap;
import org.nutz.log.Log;
import org.nutz.log.Logs;

import java.util.Arrays;
import java.util.List;

/**
 * @author wizzer.cn
 */
@IocBean
public class JellyfinHandler {
    private final static Log log = Logs.get();
    private String domain = "";
    private String key = "";
    private String media = "";
    private int processCount = 0;
    private int skipCount = 0;
    private int time = 0;
    private String userId = "";

    public void init() {
        domain = System.getenv("URL");
        key = System.getenv("KEY");
        media = System.getenv("MEDIA");
        time = Integer.parseInt(Strings.sNull(System.getenv("TIME"), "3600"));
        log.infof("扫描媒体库: %s", media == null ? "全部" : media);
        if (Strings.isBlank(domain)) {
            log.info("服务器地址未设置,使用默认地址 http://127.0.0.1:8096");
            domain = "http://127.0.0.1:8096";
        } else {
            log.infof("服务器地址 %s", domain);
        }
        if (Strings.isBlank(key)) {
            log.error("API KEY未设置");
            return;
        }
        Tasks.scheduleAtFixedRate(this::run, time);
    }

    public void run() {
        this.userId = this.getUserId();
        if (Strings.isBlank(userId)) {
            log.error("未获取到管理员用户");
            return;
        }
        List<NutMap> views = this.getViews();
        List<String> medias = Arrays.asList(Strings.splitIgnoreBlank(Strings.sNull(media), ","));
        for (NutMap view : views) {
            if (medias.size() == 0 || medias.contains(view.getString("Name"))) {
                this.processedView(view);
            }
        }
    }

    public String getUserId() {
        NutMap nutMap = JellyfinUtil.getUsers(domain, key);
        if (nutMap == null) {
            log.error("未获取到用户列表");
            return null;
        }
        List<NutMap> userList = nutMap.getList("Users", NutMap.class);
        for (NutMap map : userList) {
            if (map.getAs("Policy", NutMap.class).getBoolean("IsAdministrator")) {
                return map.getString("Id");
            }
        }
        return null;
    }

    public List<NutMap> getViews() {
        NutMap nutMap = JellyfinUtil.getViews(domain, key, userId);
        if (nutMap == null) {
            log.error("未获取到媒体库");
            return null;
        }
        return nutMap.getList("Items", NutMap.class);
    }

    public void renderItems(List<NutMap> items) {
        List<String> folders = Arrays.asList("Folder", "CollectionFolder");
        List<String> objects = Arrays.asList("Series", "Movie", "BoxSet", "Audio", "MusicAlbum", "MusicArtist", "Video", "Photo");
        for (NutMap item : items) {
            if (folders.contains(item.getString("Type"))) {
                this.renderFolder(item.getString("Id"), item.getString("CollectionType"));
            } else if (objects.contains(item.getString("Type"))) {
                NutMap itemDetail = JellyfinUtil.getItem(domain, key, userId, item.getString("Id"));
                if (itemDetail == null) {
                    log.error("服务器出错,请重启Jellyfin");
                    return;
                }
                if (Strings.isNotBlank(itemDetail.getString("ForcedSortName"))) {
                    log.infof("跳过 %s", itemDetail.getString("Name"));
                    this.skipCount++;
                } else {
                    log.infof("%s", itemDetail.getString("Name"));
                    String pinyin = PinyinUtil.getPingYin(itemDetail.getString("Name"));
                    itemDetail.setv("ForcedSortName", pinyin);
                    JellyfinUtil.postItem(domain, key, item.getString("Id"), itemDetail);
                    this.processCount++;
                }
            } else {
                log.infof("跳过，未知类型：%s", Json.toJson(item));
                this.skipCount++;
            }
        }
    }

    public void renderFolder(String viewId, String collectionType) {
        if ("music".equalsIgnoreCase(collectionType)) {
            // 专辑
            NutMap musicItems = JellyfinUtil.getMusicItems(domain, key, userId, viewId);
            if (musicItems != null) {
                this.renderItems(musicItems.getAsList("Items", NutMap.class));
            }
            //艺术家
            NutMap artists = JellyfinUtil.getArtists(domain, key, userId, viewId);
            if (artists != null) {
                this.renderItems(artists.getAsList("Items", NutMap.class));
            }
        } else {
            NutMap nutMap = JellyfinUtil.getItems(domain, key, userId, viewId);
            if (nutMap != null) {
                this.renderItems(nutMap.getAsList("Items", NutMap.class));
            }
        }
    }

    public void processedView(NutMap view) {
        this.initCount();
        long a = System.currentTimeMillis();
        log.infof("开始处理 %s", view.getString("Name"));
        this.renderFolder(view.getString("Id"), view.getString("CollectionType"));
        log.infof("已跳过：%d，已处理：%d, 耗时 %d ms", this.skipCount, this.processCount, (System.currentTimeMillis() - a));
    }

    private void initCount() {
        this.processCount = 0;
        this.skipCount = 0;
    }

}

File: src/JellyfinHandler.py
import os
import json
import time
import threading
from PinyinUtil import PinyinUtil  # 需要实现PinyinUtil
from JellyfinUtil import JellyfinUtil  # 需要实现JellyfinUtil
import logging

class JellyfinHandler:
    def __init__(self):
        self.log = logging.getLogger(__name__)
        self.domain = ""    #http://10.10.10.10:8096
        self.key = ""       #API KEY
        self.media = ""     #需要处理的媒体库
        self.process_count = 0
        self.skip_count = 0
        self.time = 0
        self.user_id = ""

    def init(self):
        self.domain = os.getenv("URL", "http://127.0.0.1:8096") # 服务器地址
        self.key = os.getenv("KEY")                             # API KEY
        self.media = os.getenv("MEDIA", "")                     # 需要处理的媒体库
        self.time = int(os.getenv("TIME", "3600"))              # 扫描间隔时间
        self.log.info(f"扫描媒体库: {self.media if self.media else '全部'}")
        
        if not self.domain:
            self.log.info("服务器地址未设置,使用默认地址 http://127.0.0.1:8096")
        else:
            self.log.info(f"服务器地址 {self.domain}")
        
        if not self.key:
            self.log.error("API KEY未设置")
            return
        
        # 延迟10秒执行,防止重启后Jellyfin服务未启动完成
        threading.Timer(10, self.run).start()
        threading.Timer(self.time, self.run).start()

    def run(self):
        self.user_id = self.get_user_id()
        if not self.user_id:
            self.log.error("未获取到管理员用户")
            return
        
        views = self.get_views()
        medias = self.media.split(",") if self.media else []
        for view in views:
            if not medias or view.get("Name") in medias:
                self.processed_view(view)

    def get_user_id(self):
        nut_map = JellyfinUtil.get_users(self.domain, self.key)
        if not nut_map:
            return None
        
        user_list = nut_map.get("Users", [])
        for user in user_list:
            if user.get("Policy", {}).get("IsAdministrator"):
                return user.get("Id")
        return None

    def get_views(self):
        nut_map = JellyfinUtil.get_views(self.domain, self.key, self.user_id)
        if not nut_map:
            self.log.error("未获取到媒体库")
            return []
        
        return nut_map.get("Items", [])

    def render_items(self, items):
        folders = ["Folder", "Season", "CollectionFolder"]
        objects = ["Series", "Movie", "BoxSet", "Audio", "MusicAlbum", "MusicArtist", "Video", "Photo", "Episode"]
        
        for item in items:
            if item.get("Type") in folders:
                self.render_folder(item.get("Id"), item.get("CollectionType"))
            elif item.get("Type") in objects:
                item_detail = JellyfinUtil.get_item(self.domain, self.key, self.user_id, item.get("Id"))
                if not item_detail:
                    self.log.error("服务器出错,请重启Jellyfin")
                    return
                
                forced_sort_name = item_detail.get("ForcedSortName")
                pinyin = PinyinUtil.get_pingyin(item_detail.get("Name", "")) if item_detail.get("Name") else ""
                if pinyin and len(pinyin) > 50:
                    pinyin = pinyin[:50]
                
                if forced_sort_name and pinyin == forced_sort_name:
                    self.log.info(f"跳过 {item_detail.get('Name')}")
                    self.skip_count += 1
                else:
                    self.log.info(f"处理 {item_detail.get('Name')}")
                    item_detail["ForcedSortName"] = pinyin
                    JellyfinUtil.post_item(self.domain, self.key, item.get("Id"), item_detail)
                    self.process_count += 1
            else:
                self.log.info(f"跳过，未知类型：{json.dumps(item)}")
                self.skip_count += 1

    def render_folder(self, view_id, collection_type):
        if collection_type is not None and collection_type.lower() == "music":
            # 专辑
            music_items = JellyfinUtil.get_music_items(self.domain, self.key, self.user_id, view_id)
            if music_items:
                self.render_items(music_items.get("Items", []))
            
            # 艺术家
            artists = JellyfinUtil.get_artists(self.domain, self.key, self.user_id, view_id)
            if artists:
                self.render_items(artists.get("Items", []))
        else:
            nut_map = JellyfinUtil.get_items(self.domain, self.key, self.user_id, view_id)
            if nut_map:
                self.render_items(nut_map.get("Items", []))

    def processed_view(self, view):
        self.init_count()
        start_time = time.time()
        self.log.info(f"开始处理 {view.get('Name')}")
        self.render_folder(view.get("Id"), view.get("CollectionType"))
        elapsed_time = time.time() - start_time
        self.log.info(f"已跳过：{self.skip_count}，已处理：{self.process_count}, 耗时 {elapsed_time:.2f} 秒")

    def init_count(self):
        self.process_count = 0
        self.skip_count = 0


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    handler = JellyfinHandler()
    handler.init()
